File: api/routers/congress.py
"""/api/v1/congress: the Congress overview page (chamber composition, legislative activity, and
the bills that passed both chambers).

Every figure is a mart column (docs/adr/0012, 0013); this router selects and reshapes rows and
does no arithmetic. Composition covers all 535 seats and is hand-maintained; every count and
the passed-both table cover every bill in the database, whoever sponsored it.
"""

from __future__ import annotations

from datetime import UTC, datetime
from typing import Annotated, Any

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy import text
from sqlalchemy.orm import Session

from api.db import get_session
from api.schemas.congress import (
    Activity,
    ChamberComposition,
    Composition,
    OverviewResponse,
    PartyGroup,
    PassedBoth,
    PassedBothItem,
)
from api.schemas.members import SourceRef

router = APIRouter(prefix="/congress", tags=["congress"])

MAJORITY_SQL = text(
    """
    SELECT chamber, chamber_seats, seated, vacant, congress_seats, congress_seated,
           congress_vacant, majority_threshold, tiebreak_letter, majority_pct,
           republican_caucus, democratic_caucus,
           majority_party, majority_letter, majority_margin, as_of, source, source_url, fetched_at
    FROM mart.chamber_majority
    ORDER BY CASE chamber WHEN 'house' THEN 1 ELSE 2 END
    """
)

GROUPS_SQL = text(
    """
    SELECT chamber, party_group, party_label, seats, seat_pct, caucus_with
    FROM mart.chamber_composition
    ORDER BY chamber, sort_order
    """
)

OVERVIEW_SQL = text("SELECT * FROM mart.congress_overview")

PASSED_BOTH_SQL = text(
    """
    SELECT congress, bill_type, bill_number, label, title, outcome, public_law_number,
           outcome_date, house_yea, house_nay, senate_yea, senate_nay, congress_gov_url
    FROM mart.congress_bill_outcome
    WHERE passed_both_chambers
    ORDER BY outcome_date DESC NULLS LAST, label
    """
)

ACTIVITY_FIELDS = tuple(Activity.model_fields)


def _source(row: Any) -> SourceRef:
    return SourceRef(
        source=row["source"], source_url=row["source_url"], fetched_at=row["fetched_at"]
    )


@router.get(
    "/overview",
    response_model=OverviewResponse,
    summary="Chamber composition and the tracked members' legislative activity",
)
def overview(session: Annotated[Session, Depends(get_session)]) -> OverviewResponse:
    majority = session.execute(MAJORITY_SQL).mappings().all()
    stats = session.execute(OVERVIEW_SQL).mappings().first()
    if not majority or stats is None:
        raise HTTPException(status_code=503, detail="The Congress overview marts are empty.")
    groups = session.execute(GROUPS_SQL).mappings().all()
    passed = session.execute(PASSED_BOTH_SQL).mappings().all()

    chambers = [
        ChamberComposition(
            chamber=row["chamber"],
            seats=row["chamber_seats"],
            seated=row["seated"],
            vacant=row["vacant"],
            majority_threshold=row["majority_threshold"],
            tiebreak_letter=row["tiebreak_letter"],
            majority_pct=row["majority_pct"],
            republican_caucus=row["republican_caucus"],
            democratic_caucus=row["democratic_caucus"],
            majority_party=row["majority_party"],
            majority_letter=row["majority_letter"],
            majority_margin=row["majority_margin"],
            groups=[
                PartyGroup(
                    party_group=g["party_group"],
                    label=g["party_label"],
                    seats=g["seats"],
                    seat_pct=g["seat_pct"],
                    caucus_with=g["caucus_with"],
                )
                for g in groups
                if g["chamber"] == row["chamber"]
            ],
            source_url=row["source_url"],
        )
        for row in majority
    ]
    first = majority[0]
    return OverviewResponse(
        congress=stats["congress"],
        congress_start=stats["congress_start"],
        congress_end=stats["congress_end"],
        composition=Composition(
            as_of=first["as_of"],
            seats=first["congress_seats"],
            seated=first["congress_seated"],
            vacant=first["congress_vacant"],
            chambers=chambers,
        ),
        activity=Activity(**{k: stats[k] for k in ACTIVITY_FIELDS}),
        passed_both=PassedBoth(
            total=stats["passed_both"],
            enacted=stats["passed_both_enacted"],
            adopted=stats["passed_both_adopted"],
            vetoed=stats["passed_both_vetoed"],
            items=[
                PassedBothItem(**{k: p[k] for k in PassedBothItem.model_fields}) for p in passed
            ],
        ),
        generated_at=datetime.now(UTC),
        sources=[_source(first), _source(stats)],
    )

File: site/src/lib/congress.ts
/** Model for the Congress overview page (/congress). Formatting and layout only: every count,
 *  share, threshold and margin is a mart column carried through /api/v1/congress/overview
 *  (docs/adr/0012, 0013). Nothing here adds, divides or compares figures. */
import { formatDate, formatNumber, formatShare, congressLabel } from './format';
import type {
  ChamberComposition,
  CongressOverviewResponse,
  PartyGroup,
  PassedBothItem,
} from './types';

export interface CompositionSegment {
  key: PartyGroup['party_group'];
  /** "218 R" inside the bar, or null when the segment is too narrow to hold text. */
  barLabel: string | null;
  widthPct: number;
  /** "Republican 218", "Independent 1 · caucus with R". */
  legend: string;
  title: string;
}

export interface ChamberBar {
  chamber: 'house' | 'senate';
  name: string;
  seatsLine: string;
  /** "Republicans control the House": who holds the majority, in words. */
  headline: string;
  /** Which party the headline is about, for the badge; null when nobody has a majority. */
  controlParty: 'republican' | 'democratic' | null;
  /** "219 seats caucus with Republicans (218 Republicans + 1 independent) · 214 with Democrats". */
  detail: string;
  majorityPct: number;
  majorityLabel: string;
  ariaLabel: string;
  segments: CompositionSegment[];
}

export interface StatCell {
  label: string;
  value: string;
  sub: string;
  note?: string;
}

export interface PassedRow {
  key: string;
  label: string;
  href: string;
  title: string;
  billPath: string;
  houseVote: string;
  senateVote: string;
  outcome: PassedBothItem['outcome'];
  outcomeLabel: string;
}

export interface CongressModel {
  title: string;
  dateRange: string;
  seatedLine: string;
  composition: {
    asOf: string;
    sources: { label: string; href: string }[];
    chambers: ChamberBar[];
  };
  activity: {
    heading: string;
    chip: string;
    meta: string;
    /** What the counts cover, in a sentence, with a link to the tracked members. */
    scope: string;
    trackedLabel: string;
    trackedHref: string;
    stats: StatCell[];
  };
  passedBoth: { chip: string; meta: string; scope: string; rows: PassedRow[] };
  footnote: string;
}

/** A bar segment holds its text only when it is wide enough to read; the legend carries every
 *  count regardless. This is layout, not a figure. */
const MIN_LABEL_PCT = 8;

const CAUCUS_LETTER = { republican: 'R', democratic: 'D' } as const;
const PARTY_LETTER: Record<string, string> = {
  republican: 'R',
  democratic: 'D',
  independent: 'I',
};

function segment(g: PartyGroup): CompositionSegment {
  const caucus = g.caucus_with ? ` · caucus with ${CAUCUS_LETTER[g.caucus_with]}` : '';
  const letter = PARTY_LETTER[g.party_group];
  return {
    key: g.party_group,
    barLabel: letter && g.seat_pct >= MIN_LABEL_PCT ? `${g.seats} ${letter}` : null,
    widthPct: g.seat_pct,
    legend: `${g.label} ${g.seats}${caucus}`,
    title: `${g.label}: ${g.seats} seats${g.caucus_with ? `, caucuses with ${g.caucus_with}` : ''}`,
  };
}

const PARTY_NOUN = { republican: 'Republicans', democratic: 'Democrats' } as const;
type Party = keyof typeof PARTY_NOUN;

/* One side's seat count in words. The party's own seats and any independents who caucus with
   it are both mart columns, so nothing is added here: "219 (218 Republicans + 1 independent)". */
function caucusSide(c: ChamberComposition, party: Party, total: number, short = false): string {
  const own = c.groups.find((g) => g.party_group === party);
  const indep = c.groups.find((g) => g.party_group === 'independent' && g.caucus_with === party);
  const detail =
    indep && own
      ? ` (${own.seats} ${PARTY_NOUN[party]} + ${indep.seats} independent${indep.seats === 1 ? '' : 's'})`
      : '';
  return `${total} ${short ? 'with' : 'seats caucus with'} ${PARTY_NOUN[party]}${detail}`;
}

function chamberBar(c: ChamberComposition): ChamberBar {
  const name = c.chamber === 'house' ? 'House' : 'Senate';
  const party: Party | null =
    c.majority_party === 'republican' || c.majority_party === 'democratic' ? c.majority_party : null;
  const rep = caucusSide(c, 'republican', c.republican_caucus);
  const dem = caucusSide(c, 'democratic', c.democratic_caucus);
  const repShort = caucusSide(c, 'republican', c.republican_caucus, true);
  const demShort = caucusSide(c, 'democratic', c.democratic_caucus, true);
  return {
    chamber: c.chamber,
    name,
    seatsLine: c.tiebreak_letter
      ? `${c.seats} seats · ${c.tiebreak_letter} Vice President breaks ties`
      : `${c.seats} seats`,
    headline: party
      ? `${PARTY_NOUN[party]} control the ${name}`
      : `Neither party has a majority in the ${name}`,
    controlParty: party,
    detail: party === 'republican' ? `${rep} · ${demShort}` : party === 'democratic' ? `${dem} · ${repShort}` : `${rep} · ${demShort}`,
    majorityPct: c.majority_pct,
    majorityLabel: `${c.majority_threshold} seats for a majority`,
    ariaLabel: `${name} composition: ${c.groups.map((g) => `${g.label} ${g.seats}`).join(', ')}. ${c.majority_threshold} seats are needed for a majority.`,
    segments: c.groups.map(segment),
  };
}

function vote(yea: number | null, nay: number | null): string {
  return yea === null || nay === null ? 'No roll call recorded' : `${yea}–${nay}`;
}

const OUTCOME_LABEL = {
  law: 'Law',
  vetoed: 'Vetoed',
  overridden: 'Overridden',
  adopted: 'Adopted',
  pending: 'Not enacted yet',
} as const;

function passedRow(item: PassedBothItem): PassedRow {
  const base = `${item.congress}/${item.bill_type}/${item.bill_number}`;
  const law = item.outcome === 'law' && item.public_law_number;
  return {
    key: base,
    label: item.label,
    href: item.congress_gov_url,
    title: item.title ?? item.label,
    billPath: `/bills/${base}`,
    houseVote: vote(item.house_yea, item.house_nay),
    senateVote: vote(item.senate_yea, item.senate_nay),
    outcome: item.outcome,
    outcomeLabel: law
      ? `Law ${item.public_law_number}`
      : item.outcome
        ? OUTCOME_LABEL[item.outcome]
        : 'Passed both',
  };
}

const plural = (n: number, one: string, many: string) => `${formatNumber(n)} ${n === 1 ? one : many}`;

export function buildCongressModel(data: CongressOverviewResponse): CongressModel {
  const a = data.activity;
  const { composition } = data;
  const sources = composition.chambers.map((c) => ({
    label: c.chamber === 'house' ? 'Clerk of the House' : 'Senate.gov',
    href: c.source_url,
  }));
  const asOf = formatDate(composition.as_of);
  const pb = data.passed_both;

  return {
    title: `The ${congressLabel(data.congress)}`,
    dateRange: `${formatDate(data.congress_start)} — ${formatDate(data.congress_end)}`,
    seatedLine: `Seated ${formatNumber(composition.seated)} of ${formatNumber(composition.seats)} · ${plural(composition.vacant, 'vacancy', 'vacancies')}`,
    composition: { asOf, sources, chambers: composition.chambers.map(chamberBar) },
    activity: {
      heading: 'Legislative activity',
      chip: 'All sponsors',
      meta: `${congressLabel(data.congress)} to date`,
      scope: `Every bill in our database (${formatNumber(a.bills_in_dataset)}), whoever sponsored it: bills our ${a.tracked_members} tracked members sponsored or cosponsored, plus any bill a recorded roll call named. Not yet every bill in Congress.`,
      trackedLabel: 'See tracked members',
      trackedHref: '/members',
      stats: [
        {
          label: 'Bills in our database',
          value: formatNumber(a.bills_in_dataset),
          sub: `${formatNumber(a.bills_house)} House · ${formatNumber(a.bills_senate)} Senate`,
        },
        {
          label: 'Passed a chamber',
          value: formatNumber(a.passed_chamber),
          sub: `${formatNumber(a.passed_chamber_house_origin)} House bills · ${formatNumber(a.passed_chamber_senate_origin)} Senate bills`,
        },
        {
          label: 'Became law',
          value: formatNumber(a.became_law),
          sub: `${formatShare(a.became_law_pct)} of bills in our database`,
        },
        {
          label: 'Vetoed',
          value: formatNumber(a.vetoed),
          sub: `${formatNumber(a.vetoed_overridden)} overridden · ${formatNumber(a.vetoed_not_overridden)} not overridden`,
        },
      ],
    },
    passedBoth: {
      chip: 'All sponsors',
      scope: 'The same bills as the counts above.',
      meta: `${plural(pb.total, 'measure', 'measures')} · ${formatNumber(pb.enacted)} enacted · ${formatNumber(pb.adopted)} adopted · ${formatNumber(pb.vetoed)} vetoed`,
      rows: pb.items.map(passedRow),
    },
    footnote: `Composition figures hand-maintained, as of ${asOf} · bill and vote figures from the Congress.gov API and Senate.gov roll calls, refreshed nightly`,
  };
}
